### PR TITLE
Add Mid documentation

### PR DESCRIPTION
**Pull request by mikhailchuryakov**
translate Oleg's [note](https://gist.github.com/Odomontois/18503ae41f9b283227dbee727a627468)

### CODE REVIEW

File: docs/mid.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+Consider some trait
+```scala
+trait MyBusinessModule[F[_]]{
+  def doBusinessThing(entity: Entity, info: Info): F[Value]
+  def undoBusinessThing(entity: Entity): F[Respect]
+}
+```
+Often `F` presented like some `IO` or any transformer
+
+But signature doesn't oblige to be strict. Moreover, there are no necessity to use functor
```
Comment (oskin1) [inline docs/mid.md:22]:
```suggestion
But signature doesn't oblige to be strict. Moreover, there is no necessity to use functor
```

File: docs/mid.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+```
+This is contravariant type. The module takes the form
+```scala
+//MyBusinessModule[Post[F, *]]
+{
+  def doBusinessThing(entity: Entity, info: Info): Value => F[Unit]
+  def undoBusinessThing(entity: Entity): Respect => F[Unit]
+}
+```
+Having been received the result the new essence could be taken to log or, for example, send the events to kafka
```
Comment (oskin1) [inline docs/mid.md:52]: This sentence is a bit misleading 

```suggestion
Such an implementation of a module can express logging or validation of a computation result.
```

File: docs/mid.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+Applying this to the module
+```scala
+//MyBusinessModule[Mid[F, *]]
+{
+  def doBusinessThing(entity: Entity, info: Info): F[Value] => F[Value]
+  def undoBusinessThing(entity: Entity): F[Respect] => F[Respect]
+}
+```
+It is capable of what the previous ones, but also "run" `F` multiple times or not run at all
```
Comment (oskin1) [inline docs/mid.md:68]:
```suggestion
`Mid` provides capabilities of both `Pre` and `Post`, but also allows to "run" `F` multiple times or not to run it at all.
```

File: docs/mid.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+Such middleware can be caching, retrying or another logic, which is not implemented in infrastructure, but demand additional reflection
```
Comment (oskin1) [inline docs/mid.md:70]:
```suggestion
Such middleware can be caching, retrying or another logic, which is not implemented in infrastructure, but requires additional reflection
```

File: docs/mid.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+## Usage
+
+It turns out enough [ApplyK](https://typelevel.org/cats-tagless/api/cats/tagless/ApplyK.html). Via
```
Comment (oskin1) [inline docs/mid.md:74]:
```suggestion
It turns out that [ApplyK](https://typelevel.org/cats-tagless/api/cats/tagless/ApplyK.html) is enough. Via
```

File: docs/mid.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+```scala
+def map2K[F[_], G[_], H[_]](af: A[F], ag: A[G])(f: Tuple2K[F, G, *]~> H]): A[H]
+```
+Possible to compose main and plugin results i.e. origin implementation and addictive
```
Comment (oskin1) [inline docs/mid.md:78]:
```suggestion
Its makes it possible to compose the result of a main computation and the result of a plug-in computation. Hence, we can also compose main module implementation and pluggable one.
```

File: docs/mid.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+`map2K` call with `F = F, G = Mid[F, *], H = F` substitute, gives `MyBusinessModule[F]` and plugin `MyBusinessModule[Mid]`
+as `af` and `ag`, only remains to implements `Tuple2K[F, G, *]~> H]` i.e. `[A] (F[A], F[A] => F[A]) => F[A]` or `(fa, f) => f(fa)`
```
Comment (oskin1) [inline docs/mid.md:81]: This is misleading

**Comment by catostrophe** [comment]
Please check grammar with Grammarly

**Comment by mikhailchuryakov** [comment]
Sorry, forgot about Grammarly

File: docs/mid.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+## Assumption
+
+Consider some trait
+```scala
+trait MyBusinessModule[F[_]]{
+  def doBusinessThing(entity: Entity, info: Info): F[Value]
+  def undoBusinessThing(entity: Entity): F[Respect]
+}
+```
+Often `F` presented like some `IO` or any transformer
```
Comment (FunFunFine) [inline docs/mid.md:20]: Shouldn't we add the fact that `F` could some kind of reader too? I feel like this is not being told enough 😞

File: docs/mid.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+But signature doesn't oblige to be strict. Moreover, there is no necessity to use a functor
+
+Let's start with an example
+```scala
+type Pre[F[_], A] = F[Unit]
+``` 
+Despite `Pre` has type-parameter, it doesn't put any information to the result
```
Comment (FunFunFine) [inline docs/mid.md:28]:
```suggestion
Despite `Pre` has type-parameter `A`, it doesn't put any information to the result
```

File: docs/mid.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+Apply `MyBusinessModule[F[_]]` to `Pre[F[_], *]`
+```scala
+// MyBusinessModule[Pre[F, *]]
```
Comment (FunFunFine) [inline docs/mid.md:32]: Why exactly is this commented out?

File: docs/mid.md (hunk 8 above)
Comment (mikhailchuryakov) [reply docs/mid.md]: Added

File: docs/mid.md (hunk 10 above)
Comment (mikhailchuryakov) [reply docs/mid.md]: Fixed it

**Review by Odomontois — APPROVED** [review]

**Pull request merged**